### ISSUE TITLE
Class weight

### ISSUE THREAD

**Issue by Adblu** [ultralytics/yolov5#5976]
### Search before asking

- [X] I have searched the YOLOv5 [issues](https://github.com/ultralytics/yolov5/issues) and [discussions](https://github.com/ultralytics/yolov5/discussions) and found no similar questions.

### Question

I have highly imbalanced dataset. Related to this article: [link](https://androidkt.com/how-to-use-class-weight-in-crossentropyloss-for-an-imbalanced-dataset/) is it possible to do it in yolov5 ?

**Comment by glenn-jocher** [comment]
@Adblu class weights are automatically computed here. You can apply these in loss.py as you see fit.

https://github.com/ultralytics/yolov5/blob/d699c21c752a9d9ca26232fabb91d55e4daea5f8/train.py#L255

**Comment by ingbeeedd** [comment]
@glenn-jocher 

Class_weights is used for image_weight, but in loss, does not the class_weight value be considered separately because it is BCE?

**Comment by glenn-jocher** [comment]
@ingbeeedd I'm sorry I don't understand your question. Image weights attempts to selectively load a dataset. Class weights are produced for all datasets and can optionally be used in loss computation by customizing the code.

**Comment by uysalfurkan** [comment]
@glenn-jocher How to customize the code? Could you give an example?

**Comment by glenn-jocher** [comment]
@uysalfurkan To get started, you can modify the `compute_loss` function to use class weights in place of the default class BCE loss. Specifically, you can multiply the output of the main loss function by the class weights before backward pass.

```
def compute_loss(pred, targets):
    ...
    class_weights = ...
    cls_loss = F.binary_cross_entropy_with_logits(cls_pred, cls_targets, reduction='none')
    loss = (obj_loss * obj_mask * class_weights + cls_loss * cls_mask * class_weights).sum() / nT
    return loss
``` 

Where the class weights is computed using the following function:
```
def get_class_weight(labels, c=1.02):
    n_samples = len(labels)
    n_classes = labels.max() + 1
    frequency = np.zeros(n_classes)
    for label in labels:
        frequency[label] += 1

    p = frequency / n_samples
    class_weight = 1 / (np.log(c + p))
    return class_weight
```

Here, `labels` denotes the class label distribution in the dataset. The `c` hyper-parameter can be tuned based on the severity of class imbalance.

You can modify this code to fit your specific use case.

Let me know if this helps or if you have any more questions!
